### PR TITLE
516 - Accordion Functional and E2E Tests

### PR DESCRIPTION
**Pull request by tjamesallen15**
**Explain the _details_ for making this change. What existing problem does the pull request solve?**

**Related github/jira issue (required)**:

https://github.com/infor-design/enterprise/issues/516

**Steps necessary to review your pull request (required)**:

npm run functional:ci
npm run e2e:ci

### CODE REVIEW

File: test/components/accordion/accordion-api.func-spec.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+  it('Should be able to destroy it', () => {
+    accordionObj.destroy();
+
+    expect(document.body.querySelector('.btn.hide-focus')).toBeFalsy();
+  });
+
+  it('Should be able to enable it', () => {
+    accordionObj.enable();
+    accordionObj.enable();
```
Comment (tmcconechy) [inline test/components/accordion/accordion-api.func-spec.js]: Two calls here..

File: test/components/accordion/accordion.e2e-spec.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+  it('Should accordion have panels', async () => {
+    expect(await element(by.className('accordion panel'))).toBeTruthy();
+  });
+
+  it('Should panel accordion be expanded', async () => {
+    const buttonEl = await element.all(by.tagName('button')).first();
+    await buttonEl.click();
+
+    expect(await element(by.xpath('//a[@aria-expanded="true"]'))).toBeTruthy();
```
Comment (tmcconechy) [inline test/components/accordion/accordion.e2e-spec.js]: We have not used xpath much, should be able to do this with `by.css`

File: test/components/accordion/accordion.e2e-spec.js [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+  it('Should ajax accordion be displayed', async () => {
+    expect(await element(by.className('accordion'))).toBeTruthy();
+  });
+
+  it('Should ajax accordion be expanded', async () => {
+    const buttonEl = await element.all(by.tagName('button')).first();
+    await buttonEl.click();
+
+    expect(await element(by.xpath('//a[@aria-expanded="true"]'))).toBeTruthy();
```
Comment (tmcconechy) [inline test/components/accordion/accordion.e2e-spec.js]: Same here

File: test/components/accordion/accordion.e2e-spec.js [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+  it('Should accordion be displayed', async () => {
+    expect(await element(by.className('accordion'))).toBeTruthy();
+  });
+
+  it('Should accordion can be expanded', async () => {
+    const buttonEl = await element.all(by.tagName('button')).first();
+    await buttonEl.click();
+
+    expect(await element(by.xpath('//a[@aria-expanded="true"]'))).toBeTruthy();
```
Comment (tmcconechy) [inline test/components/accordion/accordion.e2e-spec.js]: And here

File: test/components/accordion/accordion-api.func-spec.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
+import { Accordion } from '../../../src/components/accordion/accordion';
+
+const accordionHTML = require('../../../app/views/components/accordion/example-index.html');
+const svg = require('../../../src/components/icons/svg.html');
+
+let accordionEl;
+let svgEl;
+let accordionObj;
+
+describe('Accordion API', () => {
```
Comment (EdwardCoyle) [inline test/components/accordion/accordion-api.func-spec.js]: Functional test suite looks pretty good to start, but there are a lot of API methods missing.  Would like to see:

- explicit testing of `expand()`, `collapse()` that check not just a single headers, but multiple headers.  Don't forget there's also an API setting `allowOnePane` that can be false, making multiple accordion panes open at the same time.  Will also need tests for that
- check the contents of any accordion using `toData()`, to make sure the object representation is correct
- use `select()` on an element and then make sure it gets the proper CSS classes to identify that it's been selected.
- `isDisabled()`, `isFiltered()`, `isExpanded()`, etc are all methods that check individual accordion headers' state.  Check all of these against various types of accordions.

The above are just some ideas, but Accordion's one of our bigger components.  You should look through the source code, find all the methods that are not labeled as `@private`, and make sure there is at least one or several good tests for each, depending on what the method does.

File: test/components/accordion/accordion.e2e-spec.js [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+  it('Should accordion be disabled', async () => {
+    expect(await element(by.className('is-disabled'))).toBeTruthy();
+  });
+
+  it('Should disabled accordion be displayed', async () => {
+    expect(await element(by.className('accordion is-disabled'))).toBeTruthy();
+  });
+});
+
+describe('Accordion example-index tests', () => {
```
[the remark this answers is not in the document]
Comment (d-g-h) [reply test/components/accordion/accordion.e2e-spec.js]: @LarryAmisola, documentation about the process is here, https://github.com/infor-design/enterprise/blob/master/docs/TESTING.md#working-with-visual-regression-tests. If you run into trouble, I can add the screenshots from the Docker container, or we can both walk through the process together

**Comment by clepore** [comment]
@tjamesallen15 Are you still working on this at all? It's been a while since we've seen any updates or pushes.

File: test/components/accordion/accordion.e2e-spec.js [hunk 7]
```diff
@@ -0,0 +1,118 @@
+const { browserStackErrorReporter } = requireHelper('browserstack-error-reporter');
+const utils = requireHelper('e2e-utils');
+
+requireHelper('rejection');
+
+jasmine.getEnv().addReporter(browserStackErrorReporter);
+
+describe('Accordion example-accordion-click-event tests', () => {
```
[the remark this answers is not in the document]
Comment (tjamesallen15) [reply test/components/accordion/accordion.e2e-spec.js]: @EdwardCoyle copy!

**Comment by tjamesallen15** [comment]
aaaa

No error on my local but the error occurs intermittently

**Pull request merged**